### DISCUSSION TITLE
Suggestions to distribute a library crate that uses diesel to manage a database

### DISCUSSION THREAD

**Discussion by Datron** [diesel-rs/diesel#4001] [Q&A] [answered]
Hello!

I'm working on a project that has a few library crates linked to a binary crate in the same repository. The library crates use diesel and store data in postgres. I'm able to run the binary crate without any issues, but when I try to import the library crates into another repository they start to throw weird errors that don't show up when I run the binary crate. An example is:

```
 the trait bound `schema::sql_types::ExperimentStatusType: QueryId` is not satisfied
    --> path/superposition/crates/experimentation_platform/src/api/experiments/helpers.rs:186:15
     |
186  |         .load(conn)?;
     |          ---- ^^^^ the trait `QueryId` is not implemented for `schema::sql_types::ExperimentStatusType`
     |          |
     |          required by a bound introduced by this call
     |
     = help: the following other types implement trait `QueryId`:
               Box<T>
               DeleteStatement<T, U, Ret>
               FromClause<F>
               query_builder::select_clause::DefaultSelectClause<QS>
               BoxedSqlQuery<'_, DB, Query>
               query_builder::where_clause::BoxedWhereClause<'a, DB>
               query_source::joins::Join<Left, Right, Kind>
               diesel::expression::ops::numeric::Add<Lhs, Rhs>
             and 674 others
     = note: required for `diesel::expression::bound::Bound<schema::sql_types::ExperimentStatusType, models::ExperimentStatusType>` to implement `QueryId`
     = note: 8 redundant requirements hidden
     = note: required for `SelectStatement<FromClause<table>, DefaultSelectClause<FromClause<table>>, ..., ...>` to implement `QueryId`
     = note: the full type name has been written to 'path/context-aware-config/target/debug/deps/experimentation_platform-98f3466be937ed14.long-type-6488301493617960843.txt'
     = note: required for `SelectStatement<FromClause<table>, DefaultSelectClause<FromClause<table>>, ..., ...>` to implement `LoadQuery<'_, _, _>`
     = note: the full type name has been written to 'path/context-aware-config/target/debug/deps/experimentation_platform-98f3466be937ed14.long-type-6488301493617960843.txt'
note: required by a bound in `diesel::RunQueryDsl::load`
    --> path/.cargo/registry/src/index.crates.io-6f17d22bba15001f/diesel-2.1.6/src/query_dsl/mod.rs:1542:15
     |
1540 |     fn load<'query, U>(self, conn: &mut Conn) -> QueryResult<Vec<U>>
     |        ---- required by a bound in this associated function
1541 |     where
1542 |         Self: LoadQuery<'query, Conn, U>,
     |               ^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `RunQueryDsl::load`

For more information about this error, try `rustc --explain E0277`.
```

I tried implementing QueryId but this error still did not go away.

I'm wondering how I can go about properly exporting the library crates so I don't get this issue. 

Also, if there is a way to export the migrations folder too so that a downstream binary crate can run it, please let me know.

**Comment by weiznich** [comment]
I fear it's not possible to help you based on the provided information. You need to provide a reproducible example for your problem so that someone can else can see what's the problem there. Otherwise you can try to use `cargo tree` to double check that you only have one version of diesel in your dependency tree.

**Reply by Datron** [reply]
Hi @weiznich 
The repository that contains both library crates and binary crates is here:
https://github.com/juspay/superposition

Here's a reproducible example of my problem. I see the same issue and there is no diesel dependency added yet. 
https://github.com/Datron/superposition_example

**Reply by weiznich** [reply]
I think that might be a bug/breaking change in `diesel-derive-enum`. At least I can reproduce it with the following commands:

```rust
git clone https://github.com/juspay/superposition
cd superposition/crates/experimentation_platform
cargo check # works fine
cargo update -p diesel-derive-enum
cargo check # fails
```

It seems like they used to generate the `QueryId` impl for custom sql types for you but stopped doing that with the 2.1 release. You can likely fix that in your code by just adding `diesel::query_builder::QueryId` to the `custom_type_derives` [as documented here](http://diesel.rs/guides/configuring-diesel-cli.html). Diesel usually generates a `diesel.toml` file that already contains this line for new projects.

**Reply by Datron** [reply]
@weiznich Ah okay. Let me try this and get back to you.

**Reply by Datron** [reply]
@weiznich I'm using diesel-derive-enum=2.0.1 for all my crates, yet I still get the error I had mentioned earlier about the trait QueryId not being implemented for the enum ExperimentStatusType. Anything else I can do?

**Reply by Datron** [reply, marked as answer]
@weiznich you can ignore the above message, the error goes away when I add these dependencies in the crate Cargo.toml file. Earlier, I was adding them to the workspace Cargo.toml file.
